Update V3 scripts in plutus-preprocessor to use the new `Data` api

zliu41 opened the pull request:
It's mostly a drop-in replacement. This reduces the compilation time of `Cardano.Ledger.Plutus.Preprocessor.Binary.V3` on my machine from 3min to 30sec.

Two of the scripts aren't updated due to incompleteness of the api in 1.42.

If you are happy with this I'll do the same to V1 and V2.

cc @ana-pantilie @Unisay

zliu41 committed:
Update V3 scripts in plutus-preprocessor to use the new `Data` api

zliu41 commented:
~`purposeIsWellformedNoDatumQ` isn't updated because data-encoded `MintValue` didn't make it into 1.42.
`datumIsWellformedQ` isn't updated due to lack of `elems` or `filter` in `PlutusTx.Data.AssocMap`.~

These are resolved in plutus 1.44

ana-pantilie commented on libs/plutus-preprocessor/src/Cardano/Ledger/Plutus/Preprocessor/Source/V3.hs:
```diff
@@ -155,8 +157,8 @@ purposeIsWellformedWithDatumQ =
         case unsafeFromBuiltinData arg of
-          PV3.ScriptContext txInfo _redeemer (PV3.SpendingScript txOutRef (Just _)) ->
-            not $ null $ P.filter ((txOutRef P.==) . PV3.txInInfoOutRef) $ PV3.txInfoInputs txInfo
+          PV3D.ScriptContext txInfo _redeemer (PV3D.SpendingScript txOutRef (Just _)) ->
+            not $ PLD.null $ PLD.filter ((txOutRef P.==) . PV3D.txInInfoOutRef) $ PV3D.txInfoInputs txInfo
```
Would matching on the `txInfo` and the `txInInfoOutRef` improve the cost here? I remember that this had a large impact on the "newer" version of `AsData` (which we eventually discarded), but I can't remember if it has any effect on the current version of `AsData`.

zliu41 replied:
Perhaps, but is optimizing these scripts particularly important? @lehins

lehins replied:
No, these are tests scripts, so they don't need to be fast. As long as they can fit within reasonable limits for per transaction execution limits we are fine

ana-pantilie commented on libs/plutus-preprocessor/src/Cardano/Ledger/Plutus/Preprocessor/Source/V3.hs:
```diff
@@ -180,9 +182,9 @@ inputsOutputsAreNotEmptyNoDatumQ =
-          PV3.ScriptContext txInfo _redeemer _scriptPurpose ->
-            not $ null (PV3.txInfoInputs txInfo) || null (PV3.txInfoOutputs txInfo)
+          PV3D.ScriptContext _txInfo _redeemer (PV3D.SpendingScript _txOutRef (Just _)) -> False
+          PV3D.ScriptContext txInfo _redeemer _scriptPurpose ->
+            not $ PLD.null (PV3D.txInfoInputs txInfo) || PLD.null (PV3D.txInfoOutputs txInfo)
```
Same q as above

ana-pantilie commented on libs/plutus-preprocessor/src/Cardano/Ledger/Plutus/Preprocessor/Source/V3.hs:
```diff
@@ -192,7 +194,7 @@ inputsOutputsAreNotEmptyWithDatumQ =
-          PV3.ScriptContext txInfo _redeemer _scriptPurpose ->
-            not $ null (PV3.txInfoInputs txInfo) || null (PV3.txInfoOutputs txInfo)
+          PV3D.ScriptContext _txInfo _redeemer (PV3D.SpendingScript _txOutRef Nothing) -> False
+          PV3D.ScriptContext txInfo _redeemer _scriptPurpose ->
+            not $ PLD.null (PV3D.txInfoInputs txInfo) || PLD.null (PV3D.txInfoOutputs txInfo)
```
Same q as above

lehins replied:
Same answer ;)

ana-pantilie reviewed (approved):
Nice!

lehins reviewed (approved):
Thank you for submitting this PR, I thought you forgot about it :stuck_out_tongue_winking_eye: 

It just needs a `./scripts/fourmolize.sh`

Also it looks like Plutus added a new primitive and thus a new entry to the PlutusV1 cost model `lengthOfArray-cpu-arguments`
There is a test failing on hydra because of that, but not on GithubActions, which is weird. You don't need to worry about that test failure, we'll figure it out in another PR, since hydra is not required for merges, I just wanted to make this observation.

Unisay commented:
> Also it looks like Plutus added a new primitive and thus a new entry to the PlutusV1 cost model `lengthOfArray-cpu-arguments`

This was a mistake, should be fixed by https://github.com/IntersectMBO/plutus/pull/6948

lehins commented:
> This was a mistake, should be fixed by https://github.com/IntersectMBO/plutus/pull/6948

Looks like we'll have to wait for a release of plutus with above PR before we can merge this one.

lehins commented on libs/plutus-preprocessor/plutus-preprocessor.cabal:
```diff
@@ -42,7 +42,7 @@ library
     cardano-ledger-binary:testlib,
     cardano-ledger-core,
-    plutus-ledger-api ^>=1.37,
+    plutus-ledger-api ^>=1.42,
```
Per my https://github.com/IntersectMBO/cardano-ledger/pull/4923#issuecomment-2726057036
```suggestion
    plutus-ledger-api >=1.43,
```

lehins committed:
Update lower bound for plutus-ledger-api

lehins commented:
This PR needs:
* to be fourmalized
* flakes updated
* plutus example scripts regenerated

@teodanciu if you can finish it up that would be awesome.

zliu41 commented:
I'm going to work on this once 1.44 is released which will include some further Plinth improvements.

zliu41 committed:
plutus 1.44

zliu41 commented:
@lehins updated plutus to 1.44, and done with formatting. Let me know how to do "plutus example scripts regenerated".

teodanciu committed:
Regenerate Plutus `Examples` module

teodanciu commented:
Hey @zliu41 , I rebased (we disallow merging master into feature branches) and regenerated the examples  in a new commit - had to force-push the branch, hope that's ok.

zliu41 commented:
@teodanciu thanks!

teodanciu committed:
Fix `UtxowSpec` by increasing the budget in the redeemer

teodanciu commented:
There's a few test failures - I fixed one, but there's a few left.  I will continue looking into it tomorrow

teodanciu committed:
Make slight structural changes to test V3 Plutus scripts

so that they get compiled to distinct binaries

teodanciu committed:
Use a bigger script in test that checks reference script size

so that we don't need so many txins to push the transaction size over
the allowed maximum
